ENH: Make numpy floor_divide and remainder agree with Python `//` and `%`.

charris opened the pull request:
This fixes the problems with divmod in #7224 and brings divmod for numpy float64 scalars and arrays into agreement with divmod for Python floats. Tests are added to enforce the new behavior and assures that the results for small integers scaled by powers of two are exact. 

Too keep the scalar and array math in sync, both the scalar math and loop code is based on a new `npy_divmod` function in `npy_math`.

EDIT: I changed the name to `npy_divmod`.

divmod example, Python floats

```
    In [1]: a = 78 * 6e-8

    In [2]: b = 6e-8

    In [3]: divmod(a, b)
    Out[3]: (77.0, 5.999999999999965e-08)
```

Before this commit numpy float64 gave

```
    In [4]: divmod(float64(a), float64(b))
    Out[4]: (78.0, 0.0)
```

After this commit numpy float64 gives

```
    In [4]: divmod(float64(a), float64(b))
    Out[4]: (77.0, 5.9999999999999651e-08)
```

charris committed:
ENH: Add new npy_divmod function to npy_math.

The new function is taken from the Python version of float_divmod and
computes the result of floor_division and modulus together so that they
can be kept compatible. This should also result in the '//' and '%'
operators applied to np.float64 and Python float returning the same
values.

The intent is to implement the ufuncs floor_divide and remainder using
the npy_divmod so that their results will also match those of '//' and
'%' while providing consistency between the two.

Note that npy_divmod uses fmod, which is very slow. As a result, the
floor_division and remainder functions are about 4x slower than the
previous versions based on the floor function, but should be more
accurate.

charris committed:
ENH: Make numpy floating scalars consistent with Python divmod.

The following numpy scalar floating functions are reimplemented
using the npy_divmod function.

- remainder ('%')
- floor_division ('//')
- divmod

Note that numpy scalars warn on zero division rather than raise.

divmod example, Python floats

    In [1]: a = 78 * 6e-8

    In [2]: b = 6e-8

    In [3]: divmod(a, b)
    Out[3]: (77.0, 5.999999999999965e-08)

Before this commit numpy float64 gave

    In [4]: divmod(float64(a), float64(b))
    Out[4]: (78.0, 0.0)

After this commit numpy float64 gives

    In [4]: divmod(float64(a), float64(b))
    Out[4]: (77.0, 5.9999999999999651e-08)

charris committed:
MAINT: Remove floor function code no longer needed in scalarmath.

The floor function is no longer needed in scalarmath as its use has been
replaced by the new pymodf function.

charris committed:
ENH: Make numpy ufuncs compatible with Python divmod.

The following numpy ufuncs are reimplemented using the npy_divmod
function.

- remainder ('%')
- floor_divide ('//')

Example of differences

Currently

    In [1]: a = np.array(78 * 6e-8)

    In [2]: b = np.array(6e-8)

    In [3]: a // b
    Out[3]: 77.0

    In [4]: a % b
    Out[4]: 5.9999999999999651e-08

Previously

    In [1]: a = np.array(78 * 6e-8)

    In [2]: b = np.array(6e-8)

    In [3]: a // b
    Out[3]: 78.0

    In [4]: a % b
    Out[4]: 0.0

The first agrees with the Python values for the same operation and is a
bit more accurate for the input values as represented internally.

Closes #7224.

charris committed:
TST: Add tests for '//' and '%' (floor_divide, remainder).

Add tests for some corner cases involving inf, zero, and nan.
Check that small integers are handled exactly.

charris commented:
Note that pymodf uses fmod, as does Python. That function alone takes 4x longer to run than the old code :( That seems to be the price for a little bit more accuracy.

pv commented on numpy/core/src/npymath/npy_math.c.src:
```diff
@@ -608,6 +608,54 @@ double npy_log2(double x)
+    if (div) {
+        floordiv = npy_floor(div);
+        if (div - floordiv > 0.5)
+            floordiv += 1.0;
```
0.5@c@, 1.0@c@ ? Probably doesn't matter.

charris replied:
Fixed.

pv commented on numpy/core/tests/test_umath.py:
```diff
@@ -222,6 +223,101 @@ def test_floor_division_complex(self):
+                b = np.array(sg2*6e-8, dtype=dt2)
+                div = np.floor_divide(a, b)
+                rem = np.remainder(a, b)
+                assert_allclose(div*b + rem, a, err_msg=msg)
```
assert_equal? Is this an exact relation?

charris replied:
No, this one is sensitive to roundoff error.

pv replied:
The test passes for me also with `assert_equal`.

charris replied:
If `modf` is IEEE compliant, equality should obtain, I just don't trust the libraries that much, especially with old x86 fpus and old libraries. The example itself was the one the exposed the defect in the old implementation. However, I suppose we could use equality just to see what failed...

pv commented on numpy/core/tests/test_scalarmath.py:
```diff
@@ -150,29 +156,87 @@ def test_divmod_basic(self):
-                div, rem = divmod(a, b)
+                div = self.floordiv(a, b)
+                rem = self.mod(a, b)
                 assert_allclose(div*b + rem, a, err_msg=msg)
```
assert_equal? Is this an exact relation?

charris replied:
Exact. Hmm, looks redundant with other tests of exactness, should probably be removed.

pv commented on numpy/core/tests/test_scalarmath.py:
```diff
@@ -150,29 +156,87 @@ def test_divmod_basic(self):
-                div, rem = divmod(a, b)
+                div = self.floordiv(a, b)
+                rem = self.mod(a, b)
                 assert_allclose(div*b + rem, a, err_msg=msg)
```
Exact?

charris replied:
Not this one.

pv replied:
But then I don't understand this PR. Wasn't the aim to maintain the invariant
``a == (a % b) + (a//b)*b``?

However, I note that this test does pass with `assert_equal`.

charris replied:
IEEE compliance requires is that `modf` will be correctly rounded but I don't think we can guarantee that `(a % b) + (a//b)*b` is exact, although it should probably be good to within an ULP

pv commented on numpy/core/tests/test_umath.py:
```diff
@@ -222,6 +223,101 @@ def test_floor_division_complex(self):
+                b = np.array(sg2*19, dtype=dt2)
+                div = np.floor(a/b)
+                rem = np.remainder(a, b)
+                assert_allclose(div*b + rem, a, err_msg=msg)
```
assert_array_almost_equal_nulp?

pv commented on numpy/core/tests/test_umath.py:
```diff
@@ -222,6 +223,101 @@ def test_floor_division_complex(self):
+                msg = fmt % (dt1, dt2, sg1, sg2)
+                a = np.array(sg1*71, dtype=dt1)
+                b = np.array(sg2*19, dtype=dt2)
+                div = np.floor(a/b)
```
floor_divide?

pv commented on numpy/core/tests/test_umath.py:
```diff
@@ -222,6 +223,101 @@ def test_floor_division_complex(self):
+            fb = b.astype(dt)
+            div = np.floor(fa/fb)
+            rem = np.remainder(fa, fb)
+            assert_equal(div, tgtdiv, err_msg=msg)
```
floor(a,b) or floor_divide(a,b)?
```
import math
print(math.floor(0.6/0.04), 0.6//0.04, divmod(0.6, 0.04))
#-> 15 14.0 (14.0, 0.039999999999999966)
```

pv commented:
I think it should be clarified what the relation between floor_divide and remainder is, if it is known, and reflect that in the tests in the roundoff cases, in case such invariant is known. In cases where it's known that rounding error can exist, `assert_almost_equal_nulp` should probably be used instead of `assert_allclose`, because the default tolerance of the latter is fairly lax.

charris committed:
DOC: Document that floor_divide and remainder are complementary.

The floor_divide (//) and remainder (%) functions are complementary
in the sense that a ~= (a % b) + (a // b).

charris commented:
Updated.
